fastcrypto-tbls/src/weight_reduction/mod.rs

```diff
@@ -0,0 +1,552 @@
+        let elem = self.queue.pop().unwrap();
+        let new_value = (elem.s * self.weights[elem.i] + self.c).to_integer();
+
+        self.queue.push(QueueElement {
+            s: (Ratio::from_integer(new_value + 1) - self.c)
+                / Ratio::from_integer(self.weights[elem.i]),
+            i: elem.i,
+        });
+        if (elem.i == self.r) && (self.r + 1 < self.weights.len()) {
+            self.r += 1;
+            self.queue.push(QueueElement {
+                s: (Ratio::new(1, 1) - self.c) / Ratio::from_integer(self.weights[self.r]),
+                i: self.r,
+            });
+        }
+
+        Some(elem.i)
+    }
+}
+
+fn generate_deltas(weights: &[u64], c: Ratio) -> impl Iterator<Item = usize> + '_ {
+    Generator::new(weights, c).expect("Invalid input to generate_deltas")
+}
+
+/// Calculates the head indices for the current batch.
+fn calc_indices_head(tickets_len: usize, deltas: &[usize]) -> Vec<usize> {
+    let exclude_indices: BTreeSet<usize> = deltas.iter().copied().collect();
+    (0..tickets_len)
+        .filter(|i| !exclude_indices.contains(i))
+        .collect()
+}
+
+/// Calculates the DP data structure with indices applied that are not in `delta`s.
+/// Returns None iff the DP head cannot be constructed which means
+/// that the current batch should be skipped.
+fn calc_dp_head(
+    beta: Ratio,
+    weights: &[u64],
+    max_adv_weight: u64,
+    deltas: &[usize],
+    tickets: &Tickets,
+) -> Option<DP> {
+    let dp_head = DP::new(
+        max_adv_weight,
+        calc_adv_tickets_target(beta, tickets.total() + deltas.len() as u64),
+    )
+    .unwrap();
+
+    let indices_head = calc_indices_head(tickets.as_slice().len(), deltas);
+    let dp_head = indices_head.into_iter().try_fold(dp_head, |dp, index| {
+        dp.apply(weights[index], tickets.get(index))
+    })?;
+
+    Some(dp_head)
+}
+
+/// Apply those indices to dp_head that are in `add_indices` but not in
+/// `exclude_indices`.
+fn apply(
+    weights: &[u64],
+    dp_head: &DP,
+    tickets: &Tickets,
+    adv_tickets_target: u64,
+    add_indices: &[usize],
+    exclude_indices: &[usize],
+) -> Option<DP> {
+    let dp = dp_head.make_copy(adv_tickets_target)?;
+
+    let exclude_set: BTreeSet<usize> = exclude_indices.iter().copied().collect();
+
+    let dp = add_indices
+        .iter()
+        .copied()
+        .filter(|index| !exclude_set.contains(index))
+        .try_fold(dp, |dp, index| dp.apply(weights[index], tickets.get(index)))?;
+
+    Some(dp)
+}
+
+/// Apply `deltas` to provided `tickets`. If after applying 0 or more
+/// deltas, a valid ticket assignment is found, returns true with
+/// `tickets` containing the corresponding ticket assignment.
+/// Otherwise, returns false with `tickets` containing
+/// the new ticket assignment after applying all deltas.
+/// `dp_head` must have all indices applied except those in `deltas` with data
+/// in `tickets`.
+fn process_batch_recursive(
```

Comment:
Rust has a max call depth, so we need some upper bound on the number of recursions to make sure this will not panic - do we have that?